stdlib/public/runtime/Metadata.cpp

```diff
@@ -321,14 +322,22 @@ class FunctionCacheEntry {
     }
 
     const Metadata *getResult() const {
-      auto opaqueResult = FlagsArgsAndResult[getFlags().getNumArguments() + 1];
+      auto opaqueResult = FlagsArgsAndResult[getFlags().getNumArguments() + 2];
       return reinterpret_cast<const Metadata *>(opaqueResult);
     }
 
     const void * const *getArguments() const {
       return getFlags().getNumArguments() == 0
               ? nullptr : &FlagsArgsAndResult[1];
     }
+
+    uint32_t getParameterFlags(unsigned index) const {
+      auto numArguments = getFlags().getNumArguments();
+      assert(index < numArguments);
+      auto *flags = reinterpret_cast<const uint32_t *>(
+          FlagsArgsAndResult[numArguments + 1]);
```

Comment:
Wait, are you storing a pointer in the metadata?  What is this pointer pointing to?  Is the assumption that callers of getFunctionTypeMetadata will permanently allocate a flags array?

Reply:
Currently I have it created as constant global array and passed as a pointer, I think it makes sense to copy or re-arrange it somehow when stored in the cache, I just didn’t do that part yet...

Reply:
If I wanted to avoid array allocation and just encode flags inline after each parameter, how should I go avoid storing uint32_t as void*?

Reply:
I would suggest storing it as two separate arrays in the metadata object, one after the other.

Reply:
Sorry but I don’t think that I fully understand what you mean, I am talking about void ** we pass to `swift_getFunctionTypeMetadata`, do you mean the same thing by metadata object? The reason I decided to go with this const array is because it’s easy to pass it in, but I would really rather do it as param1,flags1, etc. here the same way as 1/2/3 functions are going to be. I just don’t know how to encode uint32_t value as an element of void ** ...

Reply:
Well, you could extend it to intptr_t and then cast, which is easy enough to do in LLVM.  But I wasn't thinking about the fact that swift_getFunctionTypeMetadata has such an opaque interface.  Frankly, I'm not sure why it does, vs. taking the FunctionTypeFlags and a types (and now a flags) pointer.

Reply:
So here is how I am thinking of encoding arbitrary parameters since it’s easy to use uint32_t: represent each of the parameter’s metadata as void * with last bit set to 1 if it has flags, 0 otherwise, format is going to be paramN, flagsN?, ... if the last bit is not set consider next pointer in the array to be new parameter, otherwise it’s uint3_t representing flags for current parameter, WDYT?

Reply:
Do you mean in the interface to swift_getFunctionTypeMetadata or do you mean in the storage in a FunctionTypeMetadata?  I tend to think that separate arrays, where the flags array is optional, is a better representation overall for both cases.

Reply:
In the metadata, we can store whether there are any parameter flags in the FunctionTypeFlags.

Reply:
I mean as swift_getFunctionTypeMetadata argument, but this might be beneficial for cache too because it aways storing a bunch of zeros in the array.